### PR TITLE
[flang][cuda] Avoid intrinsics simplification in device context

### PR DESCRIPTION
**Pull request by clementval**

### CODE REVIEW

File: flang/test/Fir/CUDA/cuda-device-context.mlir [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  %35 = fir.call @_FortranASumInteger4(%31, %32, %c5_i32, %33, %34) fastmath<contract> : (!fir.box<none>, !fir.ref<i8>, i32, i32, !fir.box<none>) -> i32
+  %36 = fir.load %13 : !fir.ref<i32>
+  %37 = fir.convert %36 : (i32) -> i64
+  %38 = fir.array_coor %2(%1) %37 : (!fir.ref<!fir.array<?xi32>>, !fir.shape<1>, i64) -> !fir.ref<i32>
+  fir.store %35 to %38 : !fir.ref<i32>
+  return
+}
+
+// CHECK-LABEL: func.func @_QPsum_in_device
+// CHECK-NOT: fir.call @_FortranASumInteger4x1_contract_simplified
```
Comment (Renaud-K) [inline flang/test/Fir/CUDA/cuda-device-context.mlir:44]: Can you add a description explaining why?

Comment (clementval) [reply flang/test/Fir/CUDA/cuda-device-context.mlir:44]: Sure.

File: flang/test/Fir/CUDA/cuda-device-context.mlir [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  %34 = fir.convert %29 : (!fir.box<i1>) -> !fir.box<none>
+  %35 = fir.call @_FortranASumInteger4(%31, %32, %c5_i32, %33, %34) fastmath<contract> : (!fir.box<none>, !fir.ref<i8>, i32, i32, !fir.box<none>) -> i32
+  %36 = fir.load %13 : !fir.ref<i32>
+  %37 = fir.convert %36 : (i32) -> i64
+  %38 = fir.array_coor %2(%1) %37 : (!fir.ref<!fir.array<?xi32>>, !fir.shape<1>, i64) -> !fir.ref<i32>
+  fir.store %35 to %38 : !fir.ref<i32>
+  return
+}
+
+// Check that intrinsic simplification is disable in CUDA Fortran context. The simplified intrinsic is
```
Comment (Renaud-K) [inline flang/test/Fir/CUDA/cuda-device-context.mlir:43]: nit: "is disabled" "the choice has not been made"

Comment (clementval) [inline flang/test/Fir/CUDA/cuda-device-context.mlir:43]:
```suggestion
// Check that intrinsic simplification is disabled in CUDA Fortran context. The simplified intrinsic is
```

File: flang/test/Fir/CUDA/cuda-device-context.mlir [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  %37 = fir.convert %36 : (i32) -> i64
+  %38 = fir.array_coor %2(%1) %37 : (!fir.ref<!fir.array<?xi32>>, !fir.shape<1>, i64) -> !fir.ref<i32>
+  fir.store %35 to %38 : !fir.ref<i32>
+  return
+}
+
+// Check that intrinsic simplification is disabled in CUDA Fortran context. The simplified intrinsic is
+// created in the module op but the device func will be migrated into a gpu module op resulting in a
+// missing symbol error. 
+// The simplified intrinsic could also be migrated to the gpu module but the choice as not be made
```
Comment (clementval) [inline flang/test/Fir/CUDA/cuda-device-context.mlir:46]:
```suggestion
// The simplified intrinsic could also be migrated to the gpu module but the choice has not be made
```

File: flang/test/Fir/CUDA/cuda-device-context.mlir (hunk 2 above)
Comment (clementval) [reply flang/test/Fir/CUDA/cuda-device-context.mlir]: Thanks for catching this! Updated.

**Review by Renaud-K — APPROVED** [review]

**Pull request merged**